File: packages/chik-js/src/storage.ts
import { ChikClientError, type ChikClientOptions, type ChikCookieSessionSource, type ChikSessionTokenSource } from "./core.js";
import { ChikErrorCode, ChikErrorMessage, chikErrorCodeFromWire, isChikErrorCode, type ChikErrorCode as ChikErrorCodeValue } from "./error-contract.js";
import { chikSessionScopeHeader } from "./wire-contract.js";

export type ChikStorageClientOptions = ChikClientOptions;

export interface ChikStorageCreateUploadInput {
  bucket: string;
  key: string;
  contentType: string;
  size: number;
  contentSha256?: string | undefined;
  expiresInSeconds?: number | undefined;
  idempotencyKey?: string | undefined;
}

export interface ChikStorageDownloadInput { bucket: string; key: string; expiresInSeconds?: number | undefined; }
export interface ChikStoragePublicUrlInput { bucket: string; key: string; }

export class ChikStorageApiError extends Error {
  readonly rawCode: string | undefined;

  constructor(readonly code: ChikErrorCodeValue, message: string, readonly status: number, options: { cause?: unknown; rawCode?: string } = {}) {
    super(message, options.cause === undefined ? undefined : { cause: options.cause });
    this.name = "ChikStorageApiError";
    this.rawCode = options.rawCode;
  }
}

export class ChikStorageUploadCommitUnknownError extends Error {
  readonly code = ChikErrorCode.commitUnknown;
  constructor(readonly uploadId: string, message: string, readonly status: number, options: { cause?: unknown } = {}) {
    super(message, options.cause === undefined ? undefined : { cause: options.cause });
    this.name = "ChikStorageUploadCommitUnknownError";
  }
}

export function isChikStorageUploadCommitUnknownError(error: unknown): error is ChikStorageUploadCommitUnknownError {
  return error instanceof ChikStorageUploadCommitUnknownError;
}

export interface ChikStorageUpload {
  readonly id: string;
  readonly expiresAt: string;
  put(body: BodyInit, init?: { signal?: AbortSignal | undefined }): Promise<void>;
  complete(): Promise<unknown>;
  abort(): Promise<void>;
}

export interface ChikStorageClient {
  createUpload(input: ChikStorageCreateUploadInput): Promise<ChikStorageUpload>;
  getDownloadUrl(input: ChikStorageDownloadInput): Promise<{ url: string; expiresAt: string }>;
  getPublicUrl(input: ChikStoragePublicUrlInput): string;
}

interface SingleUpload { uploadIntentId: string; expiresAt: string; uploadUrl: string; requiredHeaders: Record<string, string>; }
interface MultipartUpload { uploadIntentId: string; expiresAt: string; uploadMode: "multipart"; partSizeBytes: string; partCount: number; multipartToken: string; }

export function createChikStorageClient(options: ChikStorageClientOptions): ChikStorageClient {
  const fetchImplementation = options.fetch ?? globalThis.fetch;
  const base = storageOrigin(options.baseUrl);
  return {
    async createUpload(input) {
      const request = {
        method: "POST",
        body: JSON.stringify({
          bucket: input.bucket,
          key: input.key,
          contentType: input.contentType,
          size: input.size,
          ...(input.contentSha256 ? { contentSha256: input.contentSha256 } : {}),
          expiresInSeconds: input.expiresInSeconds ?? 900,
          idempotencyKey: input.idempotencyKey ?? crypto.randomUUID(),
        }),
      } satisfies RequestInit;
      const authentication = storageAuthenticationRetryContext(options, request, false);
      const created = await storageJSON<SingleUpload | MultipartUpload>(fetchImplementation, base, options, "/api/storage/uploads", request, false, authentication);
      if (isMultipart(created)) return multipartUpload(fetchImplementation, base, options, input, created, authentication);
      return singleUpload(fetchImplementation, base, options, created, authentication);
    },
    getDownloadUrl(input) {
      return storageJSON(fetchImplementation, base, options, "/api/storage/download-url", { method: "POST", body: JSON.stringify({ bucket: input.bucket, key: input.key, expiresInSeconds: input.expiresInSeconds ?? 900 }) }, true);
    },
    getPublicUrl(input) {
      const url = new URL("/api/storage/public", base);
      url.searchParams.set("bucket", input.bucket);
      url.searchParams.set("key", input.key);
      return url.toString();
    },
  };
}

function singleUpload(
  fetchImplementation: typeof fetch,
  base: string,
  options: ChikStorageClientOptions,
  descriptor: SingleUpload,
  authentication: StorageAuthenticationRetryContext | undefined,
): ChikStorageUpload {
  return {
    id: descriptor.uploadIntentId,
    expiresAt: descriptor.expiresAt,
    async put(body, init = {}) {
      assertStorageAuthenticationCurrent(authentication);
      const url = sameOriginUploadURL(base, descriptor.uploadUrl);
      const signal = combinedStorageSignal(init.signal, authentication?.signal);
      let response: Response;
      try {
        response = await fetchImplementation(url, { method: "PUT", headers: descriptor.requiredHeaders, body, ...(signal ? { signal } : {}) });
      } catch (error) {
        if (authentication?.signal.aborted) throw storageAuthenticationChanged(error);
        throw error;
      }
      assertStorageAuthenticationCurrent(authentication);
      if (response.ok) return;
      const error = await storageError(response, "Storage upload");
      assertStorageAuthenticationCurrent(authentication);
      if (error.code === ChikErrorCode.commitUnknown) throw new ChikStorageUploadCommitUnknownError(descriptor.uploadIntentId, "The upload result could not be verified. Check completion before retrying.", error.status, { cause: error });
      throw error;
    },
    complete: () => storageJSON(fetchImplementation, base, options, `/api/storage/uploads/${encodeURIComponent(descriptor.uploadIntentId)}/complete`, { method: "POST" }, false, authentication),
    abort: async () => { await storageJSON(fetchImplementation, base, options, `/api/storage/uploads/${encodeURIComponent(descriptor.uploadIntentId)}`, { method: "DELETE" }, false, authentication); },
  };
}

function multipartUpload(
  fetchImplementation: typeof fetch,
  base: string,
  options: ChikStorageClientOptions,
  input: ChikStorageCreateUploadInput,
  descriptor: MultipartUpload,
  authentication: StorageAuthenticationRetryContext | undefined,
): ChikStorageUpload {
  return {
    id: descriptor.uploadIntentId,
    expiresAt: descriptor.expiresAt,
    async put(body, init = {}) {
      assertStorageAuthenticationCurrent(authentication);
      if (!(body instanceof Blob)) throw new ChikStorageApiError(ChikErrorCode.multipartBodyRequired, "Multipart uploads require a Blob or File.", 400);
      const partSize = Number(descriptor.partSizeBytes);
      if (!Number.isSafeInteger(partSize) || partSize < 1 || !Number.isSafeInteger(descriptor.partCount) || descriptor.partCount < 1 || Math.ceil(body.size / partSize) !== descriptor.partCount || body.size !== input.size) {
        throw new ChikStorageApiError(ChikErrorCode.invalidMultipartPlan, "The multipart upload plan is invalid.", 502);
      }
      const signal = combinedStorageSignal(init.signal, authentication?.signal);
      for (let part = 1; part <= descriptor.partCount; part += 1) {
        const chunk = body.slice((part - 1) * partSize, Math.min(part * partSize, body.size), input.contentType);
        await putPart(fetchImplementation, base, descriptor, part, chunk, input.contentType, signal, authentication);
      }
    },
    complete: () => storageJSON(fetchImplementation, base, options, `/api/storage/uploads/${encodeURIComponent(descriptor.uploadIntentId)}/complete`, { method: "POST" }, false, authentication),
    abort: async () => { await storageJSON(fetchImplementation, base, options, `/api/storage/uploads/${encodeURIComponent(descriptor.uploadIntentId)}`, { method: "DELETE" }, false, authentication); },
  };
}

async function putPart(
  fetchImplementation: typeof fetch,
  base: string,
  descriptor: MultipartUpload,
  part: number,
  body: Blob,
  contentType: string,
  signal: AbortSignal | undefined,
  authentication: StorageAuthenticationRetryContext | undefined,
): Promise<void> {
  const url = new URL(`/api/storage/uploads/${encodeURIComponent(descriptor.uploadIntentId)}/parts/${part}`, base);
  url.searchParams.set("token", descriptor.multipartToken);
  for (let attempt = 0; attempt < 3; attempt += 1) {
    assertStorageAuthenticationCurrent(authentication);
    let response: Response;
    try {
      response = await fetchImplementation(url, { method: "PUT", headers: { "content-type": contentType }, body, ...(signal ? { signal } : {}) });
    } catch (error) {
      if (authentication?.signal.aborted) throw storageAuthenticationChanged(error);
      throw error;
    }
    assertStorageAuthenticationCurrent(authentication);
    if (response.ok) return;
    const error = await storageError(response, "Storage upload part");
    assertStorageAuthenticationCurrent(authentication);
    if (error.code !== ChikErrorCode.commitUnknown || attempt === 2) {
      if (error.code === ChikErrorCode.commitUnknown) throw new ChikStorageUploadCommitUnknownError(descriptor.uploadIntentId, "The part upload result could not be verified. Retry the same upload.", error.status, { cause: error });
      throw error;
    }
  }
}

function isMultipart(value: SingleUpload | MultipartUpload): value is MultipartUpload { return "uploadMode" in value && value.uploadMode === "multipart"; }

async function storageJSON<Result>(
  fetchImplementation: typeof fetch,
  base: string,
  options: ChikStorageClientOptions,
  path: string,
  init: RequestInit,
  retryOnAuthenticationFailure = false,
  retry = storageAuthenticationRetryContext(options, init, retryOnAuthenticationFailure),
  retriedAfterAuthenticationFailure = false,
): Promise<Result> {
  const headers = new Headers(options.headers);
  new Headers(init.headers).forEach((value, name) => headers.set(name, value));
  assertStorageAuthenticationCurrent(retry);
  headers.set("content-type", "application/json");
  if (!headers.has("authorization")) {
    const token = retry
      ? retry.retryToken
      : (options.sessionTokenSource ? options.sessionTokenSource.getSessionToken() : options.sessionToken)
        ?? options.apiKey;
    if (token) headers.set("authorization", `Bearer ${token}`);
  }
  if (retry?.kind === "cookie") headers.set(chikSessionScopeHeader, retry.scope);
  const signal = combinedStorageSignal(init.signal, retry?.signal);
  let response: globalThis.Response;
  try {
    response = await fetchImplementation(new URL(path, base), { ...init, headers, ...(signal ? { signal } : {}), ...(headers.has("authorization") ? { redirect: "error" } : {}) });
  } catch (error) {
    if (retry?.signal.aborted) throw storageAuthenticationChanged(error);
    throw new ChikStorageApiError(ChikErrorCode.unavailable, error instanceof Error ? error.message : "Storage could not be reached.", 503, { cause: error });
  }
  assertStorageAuthenticationCurrent(retry);
  if (!response.ok) {
    if (
      response.status === 401
      && retryOnAuthenticationFailure
      && !retriedAfterAuthenticationFailure
      && retry !== undefined
    ) {
      const token = await refreshStorageAuthentication(retry);
      if (token) {
        return storageJSON(
          fetchImplementation,
          base,
          options,
          path,
          init,
          true,
          retry,
          true,
        );
      }
    }
    const error = await storageError(response, "Storage request");
    assertStorageAuthenticationCurrent(retry);
    throw error;
  }
  if (response.status === 204) return undefined as Result;
  let result: Result;
  try {
    result = await response.json() as Result;
  } catch (error) {
    if (retry?.signal.aborted) throw storageAuthenticationChanged(error);
    throw new ChikStorageApiError(ChikErrorCode.invalidResponse, "Storage returned invalid JSON.", response.status);
  }
  assertStorageAuthenticationCurrent(retry);
  return result;
}

type StorageAuthenticationRetryContext = {
  readonly scope: string;
  readonly signal: AbortSignal;
  readonly source: ChikSessionTokenSource | ChikCookieSessionSource;
  readonly kind: "token" | "cookie";
  readonly retryEnabled: boolean;
  retryToken?: string | undefined;
};

function storageAuthenticationRetryContext(
  options: ChikStorageClientOptions,
  init: RequestInit,
  enabled: boolean,
): StorageAuthenticationRetryContext | undefined {
  if (
    new Headers(options.headers).has("authorization")
    || new Headers(init.headers).has("authorization")
  ) return undefined;
  const tokenSource = options.sessionTokenSource;
  if (tokenSource) {
    const token = tokenSource.getSessionToken();
    const scope = tokenSource.getSessionScope?.();
    const signal = tokenSource.getSessionScopeSignal?.();
    if (signal?.aborted) throw storageAuthenticationChanged();
    if (scope === undefined && signal === undefined) return undefined;
    if (!scope?.trim() || !signal) throw storageAuthenticationChanged();
    return {
      kind: "token",
      source: tokenSource,
      scope,
      signal,
      retryEnabled: enabled && typeof tokenSource.refreshSession === "function",
      ...(token?.trim() ? { retryToken: token.trim() } : {}),
    };
  }
  const cookieSource = options.cookieSessionSource;
  if (!cookieSource || options.sessionToken?.trim() || options.apiKey?.trim()) return undefined;
  const scope = cookieSource.getSessionScope();
  const signal = cookieSource.getSessionScopeSignal();
  if (signal?.aborted) throw storageAuthenticationChanged();
  if (scope === undefined && signal === undefined) return undefined;
  if (!scope?.trim() || !signal) throw storageAuthenticationChanged();
  return { kind: "cookie", source: cookieSource, scope, signal, retryEnabled: enabled };
}

async function refreshStorageAuthentication(retry: StorageAuthenticationRetryContext): Promise<string | undefined> {
  const refreshSession = retry.source.refreshSession;
  if (!retry.retryEnabled || !refreshSession || !storageAuthenticationRetryScopeIsCurrent(retry)) return undefined;
  const token = await refreshSession.call(retry.source, retry.scope);
  if (!storageAuthenticationRetryScopeIsCurrent(retry)) return undefined;
  if (retry.kind === "token") {
    const normalized = token?.trim();
    if (!normalized) return undefined;
    retry.retryToken = normalized;
  }
  return retry.scope;
}

function storageAuthenticationRetryScopeIsCurrent(retry: StorageAuthenticationRetryContext): boolean {
  return !retry.signal.aborted
    && retry.source.getSessionScope?.() === retry.scope
    && retry.source.getSessionScopeSignal?.() === retry.signal;
}

function assertStorageAuthenticationCurrent(authentication: StorageAuthenticationRetryContext | undefined): void {
  if (authentication && !storageAuthenticationRetryScopeIsCurrent(authentication)) throw storageAuthenticationChanged();
}

function storageAuthenticationChanged(cause?: unknown): ChikStorageApiError {
  return new ChikStorageApiError(
    ChikErrorCode.aborted,
    ChikErrorMessage.authenticationSessionChanged,
    409,
    cause === undefined ? {} : { cause },
  );
}

function combinedStorageSignal(
  caller: AbortSignal | null | undefined,
  authentication: AbortSignal | undefined,
): AbortSignal | undefined {
  if (!caller) return authentication;
  if (!authentication) return caller;
  return AbortSignal.any([caller, authentication]);
}

async function storageError(response: Response, action: string): Promise<ChikStorageApiError> {
  const text = await response.text();
  try {
    const value = JSON.parse(text) as { code?: unknown; message?: unknown };
    const rawCode = typeof value.code === "string" && !isChikErrorCode(value.code) ? value.code : undefined;
    return new ChikStorageApiError(chikErrorCodeFromWire(value.code), typeof value.message === "string" ? value.message : `${action} failed with HTTP ${response.status}`, response.status, { rawCode });
  } catch { return new ChikStorageApiError(ChikErrorCode.unknown, `${action} failed with HTTP ${response.status}`, response.status); }
}

function storageOrigin(baseUrl: string): string {
  let url: URL;
  try { url = new URL(baseUrl); } catch { throw new ChikClientError(ChikErrorCode.invalidArgument, "baseUrl must be an absolute HTTP deployment origin.", 400); }
  if ((url.protocol !== "http:" && url.protocol !== "https:") || url.username || url.password || url.search || url.hash) throw new ChikClientError(ChikErrorCode.invalidArgument, "baseUrl must be an absolute HTTP deployment origin.", 400);
  return url.origin;
}

function sameOriginUploadURL(origin: string, value: string): string {
  try {
    const url = new URL(value);
    if (url.origin !== origin || url.username || url.password) throw new Error("different origin");
    return url.toString();
  } catch { throw new ChikStorageApiError(ChikErrorCode.invalidUploadUrl, "The upload URL is not valid for this application.", 502); }
}

File: packages/chik-dart/lib/src/storage.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import 'auth.dart';
import 'error-contract.dart';

/// HTTP error returned by the Chickinfry Storage API.
///
/// In Storage upload commit determinations, commit_unknown means the platform has not yet
/// confirmed safe completion. Do not automatically retry the same PUT or start a new upload intent.
final class ChikStorageApiError implements Exception {
  ChikStorageApiError(
    this.code,
    this.message, {
    this.rawCode,
    this.status,
    this.cause,
  });

  final ChikErrorCode code;
  final String? rawCode;
  final String message;
  final int? status;
  final Object? cause;

  @override
  String toString() => 'ChikStorageApiError($code, $message)';
}

/// Indicates Storage PUT was recorded to provider but version ACK was not received.
/// For single upload, resolve via complete() on the same upload first.
/// For multipart, put() can be called again with the same bytes; committed parts are treated as no-op.
/// Only create a new upload intent after an explicit abort().
final class ChikStorageUploadCommitUnknownError implements Exception {
  final ChikErrorCode code = ChikErrorCode.commitUnknown;
  final String uploadId;
  final String message;
  final int status;
  final Object? cause;

  ChikStorageUploadCommitUnknownError(
    this.uploadId,
    this.message, {
    this.status = 412,
    this.cause,
  });

  @override
  String toString() =>
      'ChikStorageUploadCommitUnknownError($uploadId: $message)';
}

final class ChikStorageCreateUploadInput {
  const ChikStorageCreateUploadInput({
    required this.bucket,
    required this.key,
    required this.contentType,
    required this.size,
    this.contentSha256,
    this.expiresInSeconds,
    this.idempotencyKey,
  });

  final String bucket;
  final String key;
  final String contentType;
  final int size;
  final String? contentSha256;
  final int? expiresInSeconds;
  final String? idempotencyKey;

  Map<String, Object?> toJson() => {
        'bucket': bucket,
        'key': key,
        'contentType': contentType,
        'size': size,
        if (contentSha256 != null) 'contentSha256': contentSha256,
        if (expiresInSeconds != null) 'expiresInSeconds': expiresInSeconds,
        if (idempotencyKey != null) 'idempotencyKey': idempotencyKey,
      };
}

final class ChikStorageDownloadInput {
  const ChikStorageDownloadInput({
    required this.bucket,
    required this.key,
    this.expiresInSeconds,
  });

  final String bucket;
  final String key;
  final int? expiresInSeconds;

  Map<String, Object?> toJson() => {
        'bucket': bucket,
        'key': key,
        if (expiresInSeconds != null) 'expiresInSeconds': expiresInSeconds,
      };
}

/// HTTP response abstraction used by both fetch transport and mocks.
/// Provider upload ID, signed URL, ETag, and credentials are never exposed.
final class ChikStorageResponse {
  const ChikStorageResponse(this.status, {this.bodyBytes, this.headers});

  final int status;
  final Uint8List? bodyBytes;
  final Map<String, String>? headers;

  bool get ok => status >= 200 && status < 300;

  Map<String, dynamic>? json() {
    final bytes = bodyBytes;
    if (bytes == null) return null;
    try {
      return jsonDecode(utf8.decode(bytes)) as Map<String, dynamic>;
    } catch (_) {
      return null;
    }
  }
}

/// Storage HTTP transport. The generated default uses package:http and tests
/// can replace it with canned responses.
typedef ChikStorageFetch = Future<ChikStorageResponse> Function(
  String method,
  String url, {
  Map<String, String>? headers,
  Uint8List? body,
});

/// Upload handle returned by createUpload, providing put, complete, and abort.
final class ChikStorageUpload {
  ChikStorageUpload._({
    required this.id,
    required this.expiresAt,
    required Map<String, dynamic> descriptor,
    required ChikStorageClient client,
  })  : _descriptor = descriptor,
        _client = client;

  final String id;
  final String expiresAt;
  final Map<String, dynamic> _descriptor;
  final ChikStorageClient _client;

  /// Succeeds only after both provider persistence and version record are confirmed.
  /// For multipart descriptors, splits bytes into parts and PUTs to the same origin.
  /// If commit_unknown occurs, can be recovered by calling put() on the same upload again.
  Future<void> put(Uint8List body) => _client._putUpload(_descriptor, body);

  /// Completion step invoked after single upload commit_unknown or multipart part transmission.
  Future<Map<String, dynamic>> complete() => _client._completeUpload(id);

  /// Explicitly aborts the upload intent. Create a new intent only after aborting.
  Future<void> abort() => _client._abortUpload(id);
}

final class ChikStorageClient {
  ChikStorageClient({
    required this.baseUrl,
    this.apiKey,
    this.sessionToken,
    this.sessionTokenProvider,
    this.sessionScopeProvider,
    this.sessionScopeLeaseProvider,
    this.sessionTokenRefresher,
    this.headers = const {},
    ChikStorageFetch? fetch,
  }) : _fetch = fetch ?? _chikStorageHttpFetch;

  final String baseUrl;
  final String? apiKey;
  final String? sessionToken;
  final String? Function()? sessionTokenProvider;
  final ChikSessionScopeProvider? sessionScopeProvider;
  final ChikSessionScopeLeaseProvider? sessionScopeLeaseProvider;
  final ChikSessionTokenRefresher? sessionTokenRefresher;
  final Map<String, String> headers;
  final ChikStorageFetch _fetch;
  late final String _origin = _storageOrigin(baseUrl);

  String? get _currentSessionToken {
    final token =
        sessionTokenProvider == null ? sessionToken : sessionTokenProvider!();
    final normalized = token?.trim();
    return normalized == null || normalized.isEmpty ? null : normalized;
  }

  Future<ChikStorageUpload> createUpload(
    ChikStorageCreateUploadInput input,
  ) async {
    final json = await _storageJson(
      'POST',
      '/api/storage/uploads',
      body: input.toJson(),
    );
    return ChikStorageUpload._(
      id: json['uploadIntentId'] as String,
      expiresAt: json['expiresAt'] as String,
      descriptor: json,
      client: this,
    );
  }

  Future<({String url, String expiresAt})> getDownloadUrl(
    ChikStorageDownloadInput input,
  ) async {
    final json = await _storageJson(
      'POST',
      '/api/storage/download-url',
      body: input.toJson(),
      retryOnAuthenticationFailure: true,
    );
    return (url: json['url'] as String, expiresAt: json['expiresAt'] as String);
  }

  /// Generates a stable public bucket URL synchronously without provider URLs or signatures.
  String getPublicUrl({required String bucket, required String key}) {
    final uri = Uri.parse(_origin).replace(
      path: '/api/storage/public',
      queryParameters: {'bucket': bucket, 'key': key},
    );
    return uri.toString();
  }

  Future<Map<String, dynamic>> _completeUpload(String uploadIntentId) async {
    return _storageJson(
      'POST',
      '/api/storage/uploads/${Uri.encodeComponent(uploadIntentId)}/complete',
      body: <String, Object?>{},
    );
  }

  Future<void> _abortUpload(String uploadIntentId) async {
    await _storageJson(
      'DELETE',
      '/api/storage/uploads/${Uri.encodeComponent(uploadIntentId)}',
    );
  }

  Future<void> _putUpload(
    Map<String, dynamic> descriptor,
    Uint8List body,
  ) async {
    final uploadIntentId = descriptor['uploadIntentId'] as String;
    if (descriptor['uploadMode'] == 'multipart') {
      await _putMultipart(
        descriptor: descriptor,
        uploadIntentId: uploadIntentId,
        body: body,
      );
      return;
    }
    await _putSingle(
      descriptor: descriptor,
      uploadIntentId: uploadIntentId,
      body: body,
    );
  }

  Future<void> _putSingle({
    required Map<String, dynamic> descriptor,
    required String uploadIntentId,
    required Uint8List body,
  }) async {
    final uploadUrl = _sameOriginUploadUrl(
      _origin,
      descriptor['uploadUrl'] as String,
    );
    final rawHeaders =
        descriptor['requiredHeaders'] as Map<String, dynamic>? ?? const {};
    final headers = rawHeaders.map((k, v) => MapEntry(k, v.toString()));
    final response = await _fetch(
      'PUT',
      uploadUrl,
      headers: headers,
      body: body,
    );
    if (response.ok) return;
    final error = await _storageError(response, 'Storage upload');
    if (error.code == ChikErrorCode.commitUnknown) {
      throw ChikStorageUploadCommitUnknownError(
        uploadIntentId,
        'Storage upload commit status could not be verified. Do not re-run upload.put(); verify with complete() first.',
        status: error.status ?? 502,
        cause: error,
      );
    }
    throw error;
  }

  Future<void> _putMultipart({
    required Map<String, dynamic> descriptor,
    required String uploadIntentId,
    required Uint8List body,
  }) async {
    final partSizeBytes = int.parse(descriptor['partSizeBytes'] as String);
    final partCount = descriptor['partCount'] as int;
    final token = descriptor['multipartToken'] as String;
    final contentType =
        descriptor['contentType'] as String? ?? 'application/octet-stream';
    _validateMultipartPlan(
      partSizeBytes: partSizeBytes,
      partCount: partCount,
      size: body.length,
    );
    for (var partNumber = 1; partNumber <= partCount; partNumber++) {
      final start = (partNumber - 1) * partSizeBytes;
      final end = (start + partSizeBytes > body.length)
          ? body.length
          : start + partSizeBytes;
      final part = body.sublist(start, end);
      final url = _multipartPartUrl(
        origin: _origin,
        uploadIntentId: uploadIntentId,
        partNumber: partNumber,
        token: token,
      );
      for (var attempt = 0; attempt < 3; attempt++) {
        final response = await _fetch(
          'PUT',
          url,
          headers: {'content-type': contentType},
          body: part,
        );
        if (response.ok) break;
        final error = await _storageError(
          response,
          'Storage multipart part upload',
        );
        if (error.code == ChikErrorCode.commitUnknown && attempt < 2) continue;
        if (error.code == ChikErrorCode.commitUnknown) {
          throw ChikStorageUploadCommitUnknownError(
            uploadIntentId,
            'Storage multipart part commit status could not be verified. Call upload.put() again to recover.',
            status: error.status ?? 502,
            cause: error,
          );
        }
        throw error;
      }
    }
  }

  Future<Map<String, dynamic>> _storageJson(
    String method,
    String path, {
    Map<String, Object?>? body,
    bool retryOnAuthenticationFailure = false,
  }) async {
    final encodedBody =
        body == null ? null : Uint8List.fromList(utf8.encode(jsonEncode(body)));
    final retryScope = retryOnAuthenticationFailure
        ? _captureAuthenticationRetryScope()
        : null;
    return _storageJsonEncoded(method, path, encodedBody, retryScope, false);
  }

  Future<Map<String, dynamic>> _storageJsonEncoded(
    String method,
    String path,
    Uint8List? encodedBody,
    _StorageAuthenticationRetryScope? retryScope,
    bool retriedAfterAuthenticationFailure,
  ) async {
    final requestHeaders = <String, String>{...headers};
    _storageSetHeader(requestHeaders, 'content-type', 'application/json');
    if (_storageReadHeader(requestHeaders, 'authorization') == null) {
      final token = retryScope == null
          ? _currentSessionToken ?? apiKey
          : retryScope.token;
      if (token != null && token.isNotEmpty)
        _storageSetHeader(requestHeaders, 'authorization', 'Bearer $token');
    }
    final response = await _fetch(
      method,
      '$_origin$path',
      headers: requestHeaders,
      body: encodedBody,
    );
    if (!response.ok) {
      if (!retriedAfterAuthenticationFailure &&
          response.status == 401 &&
          retryScope != null &&
          _authenticationRetryScopeIsCurrent(retryScope)) {
        final token = await sessionTokenRefresher!(
          retryScope.sessionScope,
          retryScope.lease,
        );
        if (token != null &&
            token.trim().isNotEmpty &&
            _authenticationRetryScopeIsCurrent(retryScope)) {
          retryScope.token = token.trim();
          return _storageJsonEncoded(
            method,
            path,
            encodedBody,
            retryScope,
            true,
          );
        }
      }
      throw await _storageError(response, 'Chickinfry Storage API');
    }
    if (response.status == 204) return <String, dynamic>{};
    return response.json() ?? <String, dynamic>{};
  }

  _StorageAuthenticationRetryScope? _captureAuthenticationRetryScope() {
    if (sessionTokenProvider != null &&
        sessionScopeProvider != null &&
        sessionScopeLeaseProvider != null &&
        sessionTokenRefresher != null &&
        _storageReadHeader(headers, 'authorization') == null) {
      final sessionScope = sessionScopeProvider!()?.trim();
      final lease = sessionScopeLeaseProvider!();
      final token = sessionTokenProvider!()?.trim();
      if (sessionScope != null &&
          sessionScope.isNotEmpty &&
          lease != null &&
          lease.isActive &&
          lease.sessionScope == sessionScope) {
        final scope = _StorageAuthenticationRetryScope(
          sessionScope: sessionScope,
          lease: lease,
          token: token,
        );
        return _authenticationRetryScopeIsCurrent(scope) ? scope : null;
      }
    }
    return null;
  }

  bool _authenticationRetryScopeIsCurrent(
    _StorageAuthenticationRetryScope scope,
  ) {
    return scope.lease.isActive &&
        identical(sessionScopeLeaseProvider?.call(), scope.lease) &&
        sessionScopeProvider?.call()?.trim() == scope.sessionScope;
  }

  Future<ChikStorageApiError> _storageError(
    ChikStorageResponse response,
    String action,
  ) async {
    var message = '$action failed with HTTP ${response.status}';
    var code = ChikErrorCode.unknown;
    String? unrecognizedCode;
    final body = response.json();
    if (body != null) {
      final declaredCode = body['code'];
      if (declaredCode is String && declaredCode.isNotEmpty) {
        final parsedCode = ChikErrorCode.fromWire(declaredCode);
        if (parsedCode != ChikErrorCode.unknown ||
            declaredCode == ChikErrorCode.unknown.wireValue) {
          code = parsedCode;
        } else {
          unrecognizedCode = declaredCode;
        }
      }
      final rawMessage = body['message'];
      if (rawMessage is String && rawMessage.isNotEmpty) message = rawMessage;
    }
    return ChikStorageApiError(
      code,
      message,
      rawCode: unrecognizedCode,
      status: response.status,
    );
  }
}

final class _StorageAuthenticationRetryScope {
  _StorageAuthenticationRetryScope({
    required this.sessionScope,
    required this.lease,
    required this.token,
  });

  final String sessionScope;
  final ChikSessionScopeLease lease;
  String? token;
}

void _validateMultipartPlan({
  required int partSizeBytes,
  required int partCount,
  required int size,
}) {
  if (partSizeBytes <= 0 ||
      partCount < 1 ||
      partCount > 10000 ||
      (size / partSizeBytes).ceil() != partCount) {
    throw ChikStorageApiError(
      ChikErrorCode.invalidMultipartPlan,
      'Invalid Storage multipart upload plan.',
      status: 502,
    );
  }
}

String _multipartPartUrl({
  required String origin,
  required String uploadIntentId,
  required int partNumber,
  required String token,
}) {
  if (partNumber < 1 || partNumber > 10000 || token.isEmpty) {
    throw ChikStorageApiError(
      ChikErrorCode.invalidMultipartPart,
      'Invalid Storage multipart part URL.',
      status: 502,
    );
  }
  final uri = Uri.parse(origin).replace(
    path:
        '/api/storage/uploads/${Uri.encodeComponent(uploadIntentId)}/parts/$partNumber',
    queryParameters: {'token': token},
  );
  return uri.toString();
}

String _storageOrigin(String baseUrl) {
  final uri = Uri.parse(baseUrl);
  if (uri.scheme != 'http' && uri.scheme != 'https') {
    throw ArgumentError(
      'Storage baseUrl must be an absolute HTTP(S) deployment URL.',
    );
  }
  if (uri.userInfo.isNotEmpty ||
      uri.query.isNotEmpty ||
      uri.fragment.isNotEmpty) {
    throw ArgumentError(
      'Storage baseUrl must be an absolute HTTP(S) deployment URL.',
    );
  }
  return uri.origin;
}

String _sameOriginUploadUrl(String origin, String value) {
  final uri = Uri.tryParse(value);
  if (uri == null || uri.origin != origin || uri.userInfo.isNotEmpty) {
    throw ChikStorageApiError(
      ChikErrorCode.invalidUploadUrl,
      'Storage upload URL must share the same application origin.',
      status: 502,
    );
  }
  return uri.toString();
}

final _chikStorageHttpClient = http.Client();

Future<ChikStorageResponse> _chikStorageHttpFetch(
  String method,
  String url, {
  Map<String, String>? headers,
  Uint8List? body,
}) async {
  final request = http.Request(method, Uri.parse(url))
    ..followRedirects = false
    ..maxRedirects = 0;
  if (headers != null) request.headers.addAll(headers);
  if (body != null) request.bodyBytes = body;
  final response = await http.Response.fromStream(
    await _chikStorageHttpClient.send(request),
  );
  return ChikStorageResponse(
    response.statusCode,
    bodyBytes: response.bodyBytes,
    headers: response.headers,
  );
}

String? _storageReadHeader(Map<String, String> headers, String name) {
  final lower = name.toLowerCase();
  for (final entry in headers.entries) {
    if (entry.key.toLowerCase() == lower) return entry.value;
  }
  return null;
}

void _storageSetHeader(Map<String, String> headers, String name, String value) {
  final lower = name.toLowerCase();
  headers.removeWhere((key, _) => key.toLowerCase() == lower);
  headers[name] = value;
}
